### ISSUE TITLE
zbus panics once SERIAL_NUM wraps past u32::MAX

### ISSUE THREAD

**Issue by opsound** [dbus2/zbus#946]
We have a zbus server which handles ~high dbus message rates (~thousands/s) and are seeing serial number generation cause panics once the server survives long enough to when the u32 wraps past `u32::MAX` back to zero, which triggers a panic on conversion to `NonZeroU32`

https://github.com/dbus2/zbus/blob/1cc4117f2c25c5914ea2f7452de8b1bbdea0f6a5/zbus/src/message/header.rs#L143

It's exacerbated by the fact that the reply path generates a new serial number only to overwrite it with the reply serial.
https://github.com/dbus2/zbus/blob/1cc4117f2c25c5914ea2f7452de8b1bbdea0f6a5/zbus/src/message/builder.rs#L78-L79
https://github.com/dbus2/zbus/blob/1cc4117f2c25c5914ea2f7452de8b1bbdea0f6a5/zbus/src/message/builder.rs#L40-L41
https://github.com/dbus2/zbus/blob/1cc4117f2c25c5914ea2f7452de8b1bbdea0f6a5/zbus/src/message/builder.rs#L179-L181

**Comment by zeenix** [comment]
> We have a zbus server which handles ~high dbus message rates (~thousands/s) and are seeing serial number generation cause panics once the server survives long enough to when the u32 wraps past `u32::MAX` back to zero, which triggers a panic on conversion to `NonZeroU32`

Oh no. Should have tested that better. :) I guess we can solve it simply by adding `1` each time it wraps around?

```rust
        let mut serial_num: SERIAL_NUM.fetch_add(1, SeqCst);
        if serial_num == 0 {
            serial_num = 1;
        } 

        Self {
            endian_sig: NATIVE_ENDIAN_SIG,
            msg_type,
            flags: BitFlags::empty(),
            protocol_version: 1,
            body_len,
            serial_num: serial_num.try_into().unwrap(),
        }
```

Care to provide a PR? I'm about to embark on my vacation but will be able to review simple PRs from time to time, for sure.

> It's exacerbated by the fact that the reply path generates a new serial number only to overwrite it with the reply serial.

Not sure I follow this part. We're setting the `reply_serial` field in the header, which is different from the serial number itself. The point here is to indicate which method call, the reply is for.

**Comment by opsound** [comment]
> Not sure I follow this part. We're setting the reply_serial field in the header, which is different from the serial number itself. The point here is to indicate which method call, the reply is for.

Ah - I misunderstood the code. That makes sense. I incorrectly read it as replacing the original serial that was populated as part of header generation. 

The fix you proposed sounds reasonable. It can result in there being gaps > 1 in the set of actually used serial numbers, but I think that's fine.

**Comment by opsound** [comment]
Will work on PR.

**Comment by zeenix** [comment]
> It can result in there being gaps > 1 in the set of actually used serial numbers -- is that fine?

Hmm.. how so? :thinking: 

> Will work on PR.

:+1:

**Comment by opsound** [comment]
```rust
        let mut serial_num = SERIAL_NUM.fetch_add(1, SeqCst); // A
        if serial_num == 0 { // B
            serial_num = SERIAL_NUM.fetch_add(1, SeqCst); // C
        } 
```
```
thread  | serial_num | SERIAL NUM | 
```

**Comment by opsound** [comment]
How strong of a requirement of uniqueness of serial numbers is there? What if methods are in flight while an intervening 4+ billion headers get generated, triggering the wrap around => now there's duplicate messages in flight with the same serials. Assuming this is not a realistic concern?

**Comment by zeenix** [comment]
> starting state is `SERIAL_NUM==0` (after a wrap)
> 
> * thread 0 runs A, `serial_num==0`, `SERIAL_NUM==1`
> * thread 1 runs A, `serial_num==1`, `SERIAL_NUM==2`
> * thread 0 runs B (it's true), then runs C, `serial_num==2`, `SERIAL_NUM==3`

Thanks for explaining. Note that the serial (despite its name) is not at all required to be monotonically increasing (it doesn't even have to be increasing) by the spec. It's just a good practice to treat it as such and  there is at least one specific software that's known to break from this faulty assumption but we provide a workaround for that until they fix this (see #859). 

> How strong of a requirement of uniqueness of serial numbers is there?

Only a practical one. 

> What if methods are in flight while an intervening 4+ billion headers get generated, triggering the wrap around => now there's different messages in flight with the same serials.

If this actually happens, you got much bigger issues I think. :) 

Also keep in mind that the serial number is only relevant to method calls. It doesn't mean anything for other message types, at least no meaning that the spec assigns it.

> Assuming this is not a realistic concern?

Unless, someone presents me with a use case where this actually happens and there is no easy way to workaround it, I wouldn't worry about it at all. Besides, what could possibly be the solution, especially since we want the serial numbers to be monotonically increasing?
